src/btree/row_modify.c

```diff
@@ -145,6 +145,16 @@ __wt_row_modify(WT_CURSOR_BTREE *cbt, const WT_ITEM *key, const WT_ITEM *value,
             /* If there are existing updates, append them after the new updates. */
             for (last_upd = upd; last_upd->next != NULL; last_upd = last_upd->next)
                 ;
+
+            /*
+             * If the last update on the update chain has been marked as being restored from the
+             * history store and the existing update is an in-progress prepared update, do not
+             * append the existing update chain.
+             */
+            if (F_ISSET(last_upd, WT_UPDATE_RESTORED_FROM_HS) && *upd_entry != NULL &&
+              (*upd_entry)->prepare_state == WT_PREPARE_INPROGRESS &&
+              F_ISSET(*upd_entry, WT_UPDATE_RESTORED_FROM_DS))
+                *upd_entry = NULL;
```

Comment:
Setting it to NULL leaks the memory that is allocated for the prepared updates that are allocated during the page read. You need free those instantiated prepared updates here or don't instantiate the prepared updates.

Reply:
@kommiharibabu I have pushed a change in 36b1603 that frees the update chain but I am doubtable if there is any locks required to do so.

Reply:
It feels weird to fix this in this place. Can we pass a parameter to the in_mem function to decide whether we need to instantiate the onpage prepared update or not?

Reply:
I think my proposal is hard. May be it is the only solution.

Reply:
It's not ideal to remove the update chain after instantiating it but I believe this is the first point in code path where we have knowledge of the existing and the new update chain in one place.

Reply:
Can we force to evict these pages that have prepare on it in rollback to stable to avoid this problem?

Reply:
Force evict can work for this problem, but to perform a force evict on a page, we need to track whether this page has any prepared updates and wait until all the keys on the page are processed.

Reply:
Yes, I think it is better than the current solution. We can have a flag to track whether rts removed any prepared update on the page. If it does, we force evict that page after everything is processed on that page.

Reply:
We reached here because a prepared transaction was rolled back and not through the RTS path.

Reply:
OK, but we can do the same right force evicting that page after we rollback.

Reply:
That will not solve the issue because it may still be an update restore.